In [2]:
from tensorflow.keras.preprocessing.text import one_hot

In [3]:
sent = [
    "the glass of milk",
    "the glass of juice",
    "I am learning DL",
    "I Love Marvel",
    "Loki is the Goat"
]

In [4]:
# define vocabulary size

vocab_size = 10000

In [5]:
one_hot_representation = [one_hot(word, vocab_size) for word in sent]
one_hot_representation

[[4339, 3308, 2026, 2978],
 [4339, 3308, 2026, 1322],
 [2807, 4416, 8406, 9343],
 [2807, 1582, 7757],
 [3989, 2816, 4339, 7342]]

```
so the sentence gets converted into these vectors with vocubalry size 10000

"the glass of milk" gets Converted into  [4339, 3308, 2026, 2978]  -->> these are indexes means at these indexes we will find one remaining will be zero



___
# Word Embedding Representation
___

> Now we will the indexes of sentances and pass it to Embeding Layer ,, And then we will convert it int Word Embedding Representation

In [6]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.utils import pad_sequences
from tensorflow.keras.models import Sequential

> We need to make all the sentances of fixed size

> we do that usng Pad_Sequence

# Pad Sequence

## What is Pad Sequence?

`pad_sequences()` is used to make **all sentences/sequences the same length**.

In NLP, different sentences can have different numbers of words:

```text
Sentence 1 → [4339, 3308, 2026, 2978]       → 4 words
Sentence 2 → [4339, 3308, 2026, 1322]       → 4 words
Sentence 3 → [2807, 4416, 8406, 9343]       → 4 words
Sentence 4 → [2807, 1582, 7757]             → 3 words
Sentence 5 → [3989, 2816, 4339, 7342]       → 4 words
```
> Neural networks usually need inputs with a fixed shape, so we use padding.

## Why Do We Use It?

To convert variable-length sequences into sequences of the same length

Before Padding:

[4339, 3308, 2026, 2978]
[4339, 3308, 2026, 1322]
[2807, 4416, 8406, 9343]
[2807, 1582, 7757]

              ↓
        pad_sequences()

              ↓

[4339, 3308, 2026, 2978]
[4339, 3308, 2026, 1322]
[2807, 4416, 8406, 9343]
[   0, 2807, 1582, 7757]

> 0 is used as the padding value by default.

## Syntax

```python
from tensorflow.keras.utils import pad_sequences

padded = pad_sequences(
    sequences,
    padding="pre",
    maxlen=4
)

```

- sequences → List of integer sequences
- maxlen → Maximum/fixed sequence length
- padding="pre" → Add padding at the beginning
- padding="post" → Add padding at the end


> Pad Sequence = Make all token sequences the same length so they can be fed into a neural network.

In [7]:
sent_length = 8
embedding_doc = pad_sequences(one_hot_representation, padding="pre", maxlen=sent_length)

In [8]:
embedding_doc

array([[   0,    0,    0,    0, 4339, 3308, 2026, 2978],
       [   0,    0,    0,    0, 4339, 3308, 2026, 1322],
       [   0,    0,    0,    0, 2807, 4416, 8406, 9343],
       [   0,    0,    0,    0,    0, 2807, 1582, 7757],
       [   0,    0,    0,    0, 3989, 2816, 4339, 7342]], dtype=int32)

___
# Feature representaion
___

In [9]:
# we wil use 10 dimensions
dimensions = 10

In [10]:
model = Sequential()
model.add(
    Embedding(vocab_size, dimensions, input_length=sent_length)
    )

model.compile(
    optimizer = 'adam',
    loss = "mse"
)


/home/aizen/miniconda3/envs/gensim-env/lib/python3.12/site-packages/keras/src/layers/core/embedding.py:123: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(
E0000 00:00:1788970953.760994   49338 cuda_platform.cc:52] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


In [11]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding (Embedding)           │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

In [ ]:
model.predict(embedding_doc)
# Each sentanc gets converted into 10 dimension vector

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 192ms/step


array([[[ 0.01840242, -0.04807044,  0.02316029, -0.02465728,
          0.03848464, -0.04441438,  0.04899682, -0.00134623,
          0.04754763,  0.02315838],
        [ 0.01840242, -0.04807044,  0.02316029, -0.02465728,
          0.03848464, -0.04441438,  0.04899682, -0.00134623,
          0.04754763,  0.02315838],
        [ 0.01840242, -0.04807044,  0.02316029, -0.02465728,
          0.03848464, -0.04441438,  0.04899682, -0.00134623,
          0.04754763,  0.02315838],
        [ 0.01840242, -0.04807044,  0.02316029, -0.02465728,
          0.03848464, -0.04441438,  0.04899682, -0.00134623,
          0.04754763,  0.02315838],
        [ 0.03014863, -0.01047565, -0.00153558,  0.01682016,
          0.03027815,  0.01847169, -0.04103038,  0.03374681,
         -0.04903856, -0.01888875],
        [ 0.04784067,  0.02082667, -0.01818014,  0.04159642,
         -0.04049217, -0.03664948,  0.02833431,  0.00481565,
         -0.03871479, -0.0179488 ],
        [-0.03728752,  0.01182235, -0.04148502,  0.0

In [13]:
embedding_doc[0]

array([   0,    0,    0,    0, 4339, 3308, 2026, 2978], dtype=int32)

# Embedding Layer in Simple RNN 🧠

> Internally Embedding use Word2Vec 

## What is an Embedding Layer?

An **Embedding Layer** converts integer/token IDs into **dense numerical vectors**.

```text
Token ID
   ↓
Embedding Layer
   ↓
Dense Word Vector
   ↓
Simple RNN
   ↓
Output
```

Instead of representing a word using a large one-hot vector, the Embedding Layer represents it using a smaller vector of learned numbers.

---

## Example

Suppose:

```text
the   → 4339
glass → 3308
milk  → 2026
```

The Embedding Layer may convert:

```text
4339 → [0.21, -0.45, 0.73, 0.11]
3308 → [0.82,  0.14, -0.31, 0.56]
2026 → [-0.12, 0.91, 0.33, -0.27]
```

These vectors are **learned during training**.

---

# Why Do We Use Embedding?

One-hot encoding creates large and sparse vectors:

```text
[0, 0, 0, 0, 0, 0, 1, 0, 0, ...]
```

Embedding creates a smaller and dense representation:

```text
[0.21, -0.45, 0.73, 0.11]
```

### Benefits

- Reduces dimensionality
- Creates dense representations
- Learns useful word relationships
- More efficient than large one-hot vectors
- Can capture semantic relationships

---

# Embedding Layer Syntax

```python
Embedding(
    vocab_size,
    dimensions,
    input_length=sent_length
)
```

### `vocab_size`

Number of unique words/tokens in the vocabulary.

```python
vocab_size = 10000
```

### `dimensions`

Number of values in the embedding vector for each word.

```python
dimensions = 100
```

This means every token is represented using a **100-dimensional vector**.

### `input_length`

Number of tokens in each input sequence.

If every sentence has been padded to 4 tokens:

```python
sent_length = 4
```

---

# Creating the Embedding Layer

```python
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding

model = Sequential()

model.add(
    Embedding(
        vocab_size,
        dimensions,
        input_length=sent_length
    )
)
```

---

# Embedding + Simple RNN

The Embedding Layer is **not the RNN**.

The Embedding Layer converts token IDs into vectors, and the Simple RNN processes those vectors in sequence.

```python
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN

model = Sequential()

model.add(
    Embedding(
        vocab_size,
        dimensions,
        input_length=sent_length
    )
)

model.add(
    SimpleRNN(32)
)
```

Architecture:

```text
Token IDs
    ↓
Embedding Layer
    ↓
Word Vectors
    ↓
Simple RNN
    ↓
Output
```

---

# Example with Our Padded Sequences

Suppose:

```python
[
    [4339, 3308, 2026, 2978],
    [4339, 3308, 2026, 1322],
    [2807, 4416, 8406, 9343],
    [0, 2807, 1582, 7757],
    [3989, 2816, 4339, 7342]
]
```

The Embedding Layer takes each token ID and converts it into a vector.

For example:

```text
[4339, 3308, 2026, 2978]

        ↓ Embedding

[
  vector for 4339,
  vector for 3308,
  vector for 2026,
  vector for 2978
]
```

If:

```python
dimensions = 100
```

then each token becomes a vector containing 100 numbers.

Therefore:

```text
4 tokens × 100 dimensions
        ↓
      4 × 100
```

---

# Compiling the Model

After creating the model:

```python
model.compile(
    optimizer="adam",
    loss="mse"
)
```

### `optimizer`

```python
optimizer="adam"
```

Adam controls how the model updates its weights during training.

### `loss`

```python
loss="mse"
```

MSE means **Mean Squared Error**.

> The loss function depends on the task. `mse` is commonly used for regression. Classification usually uses losses such as `binary_crossentropy` or `categorical_crossentropy`.

---

# Complete Example

```python
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN

# Vocabulary size
vocab_size = 10000

# Number of values in each word vector
dimensions = 100

# Number of tokens in each padded sentence
sent_length = 4


# Create Sequential Model
model = Sequential()


# Embedding Layer
# Converts token IDs into dense vectors
model.add(
    Embedding(
        vocab_size,
        dimensions,
        input_length=sent_length
    )
)


# Simple RNN Layer
# Processes the sequence of word vectors
model.add(
    SimpleRNN(32)
)


# Compile the model
model.compile(
    optimizer="adam",
    loss="mse"
)
```

---

# Remember

```text
Token ID
   ↓
Embedding Layer
   ↓
Dense Word Vector
   ↓
Simple RNN
   ↓
Output
```

> **Embedding = converts token IDs into learned dense vectors.**

> **Simple RNN = processes those vectors while considering the order of the words.**
